### ISSUE TITLE
Cannot use macro to define local variable(s)

### ISSUE THREAD

**Issue by arloan** [crystal-lang/crystal#9799]
main.cr:

```crystal
A = {a: 1, b: 2, c: 3}

macro def_vars
  {% for n, v in A %}
  {{n}} = {{v}}
  {% end %}
end
def_vars
puts a,b,c
```

Produces the following compiling error:
```
In main.cr:9:6

 9 | puts a,b,c
          ^
Error: undefined local variable or method 'a' for top-level

Did you mean 'p'?
exit status 1
```

Crystal version:
```
stone@LAPTOP-Q3L7M1FS:~/projects/bydb$ crystal -v
Crystal 0.35.1 [5999ae29b] (2020-06-19)

LLVM: 8.0.0
Default target: x86_64-unknown-linux-gnu
```

**Comment by asterite** [comment]
Yes, that's expected.

The reason is that parsing happens first, which determines whether something is a variable or a call. Macros are executed later.

Also, it would be quite confusing if this was the case. So I'd prefer we don't change this.

tl;dr: this is expected behavior and this is not going to change

**Comment by arloan** [comment]
> 
> 
> Yes, that's expected.
> 
> The reason is that parsing happens first, which determines whether something is a variable or a call. Macros are executed later.
> 
> Also, it would be quite confusing if this was the case. So I'd prefer we don't change this.
> 
> tl;dr: this is expected behavior and this is not going to change

Understood, thank you.

While it will be great to have an alternative way to do so in my opinion. I'll explain with codes what I want:
```crystal
...
class Foo
  getter a, b, c
  def initialize(@a : Int32, @b : String, @c : Time)
  end
end

a=1
b="hello"
c=Time.utc
foo = Foo.new(a,b,c)
...

```
As you can see, I wrote "a","b","c" many times. It will be even more tedious and frustrating if there are much more such properties/instance and local variables/arguments of same or alike names. So I decide to use macro to declare getters, i-vars and arguments, but failed in defining local variables. Hope there will be a better way.

**Comment by Blacksmoke16** [comment]
If you're only going to have getters, look into [record](https://crystal-lang.org/api/toplevel.html#record(name,*properties)-macro).

**Comment by arloan** [comment]
> 
> 
> If you're only going to have getters, look into [record](https://crystal-lang.org/api/toplevel.html#record(name,*properties)-macro).

It's close to my purpose, appreciated. While it's only for Structs, and I still need a way to declare local variables, 'cause the constructor's arguments are not just simple literals, their values come from complex business logic, and I don't like to declare those variables one by one, and don't like to write so many name again for arguments. For now I use a hash to achieve my goal, but not elegant from my view. Maybe I'm just abusing macro?

**Comment by asterite** [comment]
You can do this:

```crystal
foo = Foo.new(
  1,
  "hello",
  Time.utc,
)
```

**Comment by Blacksmoke16** [comment]
>  their values come from complex business logic

I'm still confused on how using a macro define these local variables would help?  Given you're still going to need to set the in your business logic, and pass them to `.new`.  I'm sure there is a better design/pattern you could use to make it less messy tho.  Like defining, doing some of the logic in the initializer, passing in that hash as an argument to a custom `def self.new`, or something along those lines.

**Comment by arloan** [comment]
> I'm still confused on how using a macro define these local variables would help? Given you're still going to need to set the in your business logic, and pass them to `.new`. I'm sure there is a better design/pattern you could use to make it less messy tho. Like doing some of the logic in the initializer, passing in that hash as an argument to a custom `def self.new`, or something along those lines.

Yes I can do that, wrap some business logic codes in separate functions, and cal a `.new` with a few function calls, like:
```
foo = Foobar.new(
  some_bl_fun1(arg1, arg2.. ),
  some_bl_fun2(arg3, arg4..),
  some_bl_fun3(arg5),
  ...
)
```
Just that I think it's ugly.

Using a Hash object to do initializing seems a better way, which I'm using for now, but it's still tedious 'cause I still have to write all those hash keys with the same names, those duplicated words u know, so my problem is simply not resolved.

**Comment by asterite** [comment]
You should try coding in Haskell, Elm or Elixir for a while. Then you'll know what truly verbose code means. Then come back to Crystal and you'll think "yeah, this isn't that bad, really".

**Comment by asterite** [comment]
Also, if you want more higher flexibility I recommend Ruby (I mean it). You can pass hashes to constructors, do really awesome splats, etc. Crystal being a statically typed, compiled language, doesn't provide that level of flexibility.

**Comment by Blacksmoke16** [comment]
> Yes I can do that, wrap some business logic codes in separate functions, and cal a `.new` with a few function calls, like:

That's not what I meant.  Was thinking more like:

```cr
class Foo

  def self.from_hash(hash : Hash) : self
    new hash["one"], hash["two], hash["three"]
  end

  def initialize(@one : Int32, @two : Int32, @three : Int32); end 
end

pp Foo.from_hash({"one" => 1, "two" => 2, "three" => 3})
```

**Comment by arloan** [comment]
> Also, if you want more higher flexibility I recommend Ruby (I mean it). You can pass hashes to constructors, do really awesome splats, etc. Crystal being a statically typed, compiled language, doesn't provide that level of flexibility.

Yeah, I came from Ruby, but sometimes I need to compile, and I thought Crystal can do what I want with macro. ;-p

**Comment by arloan** [comment]
> > Yes I can do that, wrap some business logic codes in separate functions, and cal a `.new` with a few function calls, like:
> 
> That's not what I meant. Was thinking more like:
> 
> ```crystal
> class Foo
>   def self.from_hash(hash : Hash) : self
>     new hash["one"], hash["two"], hash["three"]
>   end
> 
>   def initialize(@one : Int32, @two : Int32, @three : Int32); end 
> end
> 
> pp Foo.from_hash({"one" => 1, "two" => 2, "three" => 3})
> ```
> 
> This code is going to have to live somewhere, if you can abstract some of that into a method within the type you want to instantiate, that might be a good idea. However using a Hash in the first place has its own problems.

Nice idea, but we can see that u have to write those keys (one, two, three ...) for 3 times at least, which is just what I'd like to avoid.

**Comment by asterite** [comment]
Sorry, but I'm going to close this issue. Macros will never be able to do this, regardless of how much you or others want this functionality.

**Comment by rolandoam** [comment]
> Also, it would be quite confusing if this was the case. So I'd prefer we don't change this.

can you expand on why you think that's the case?

> Macros will never be able to do this, regardless of how much you or others want this functionality.

That's also a bit of a closed mentality. If *others* were a large percentage of users of the language (I have no idea if that's even remotely the case), then wouldn't you at least consider it?

I also was stuck on a similar situation, where it would be great for macros to define local variables. Here's my situation:

```crystal
macro authorized!(env)
  user = {{env}}.authorized_user?
  halt({{env}}, status_code: 400) if user.nil?
end

# using Kemal

# this won't work
get "/private/some/route" do |env|
  authorized_user!(env)
  # here I can use `user`
end

# workaround that I have to do now
get "/private/some/route" do |env|
  user = User::EMPTY
  authorized_user!(user, env)
end
```

I understand that allowing macros to define variables where they expand could be dangerous, but that's the whole point of macros. You're playing with the fabric of the language.

**Comment by straight-shoota** [comment]
The answer is basically this:

> The reason is that parsing happens first, which determines whether something is a variable or a call. Macros are executed later.

Variable declaration is handled by the parser and has an effect on parsing behaviour. The parser needs to track local variables in order to tell what's a variable and what's a call.

Maybe this could be overcome by changing how the compiler works. But that's a lot of effort and adds additional complexity.
And the benefit would be... the very questionable feature of being able to write bad code?
Having a variable appear in the current scope without explicit declaration makes it hard to read and understand what's going on.

In you example, it would be unclear where `user` is coming from if you could use it as a local variable after the macro call. 
Why not simply make this explicit?
```cr
user = authorized_user!(env)
# now you have a properly declared local variable 👍 
```

**Comment by rolandoam** [comment]
How would you return a value from a macro? At that point a function/method seems more reasonable, but then I would not be able to halt execution (`next` in the context of a Kemal route) because next inside the function does not make sense.

**Comment by oprypin** [comment]
If you have a concrete question how to best approach your particular case, please ask on the forum and I'm sure people will share some great solutions. In terms of the issue thread itself, it's not useful to bump it without new very deep analysis because the team fully agrees on the current resolution and believes it to already be the best practice.

**Comment by oprypin** [comment]
What acts as the return value of a macro is just the last expression that was present in it. But indeed, if at all possible, methods are preferred
